### ISSUE TITLE
.htaccess compatibility

### ISSUE THREAD

**Issue by Foxtrek64** [pH7Software/pH7-Social-Dating-CMS#187]
The default .htaccess file leaves the CMS impossible to install. I received repeated 500 errors from my Apache host. It seems the included file is incompatible with the latest version of Apache.

I have rewritten the .htaccess file (see below) which removes the 500 error. I have never seen an .htaccess file as long as the one in this software.

Edit: This isn't designed to be perfect. I just removed all of the items I felt to be unnecessary and merged a few sections together. I'm certain this can likely be cleaned up further.

Here is my revised htaccess file:

	<IfModule mod_rewrite.c>
		RewriteEngine On

		## Uncomment the below "#RewriteBase /" (remove "#") if pH7CMS is installed in a folder, and add the folder name after the slash "/"
		## Example: If pH7CMS is installed in "/public_html/ph7cms/", then it should be "RewriteBase /ph7cms/", however it may depends of your Apache configuration.
		## If you need, please ask your hosting company or check the Apache doc: http://httpd.apache.org/docs/current/mod/mod_rewrite.html

		RewriteBase /
		
		<IfModule mod_env.c>
			# Tell PHP that the mod_rewrite module is ENABLED.
			SetEnv HTTP_MOD_REWRITE On
		</IfModule>

		### For pH7CMS's URL router (DO NOT REMOVE THESE FOLLOWING CODE) ###
		RewriteCond %{REQUEST_FILENAME} !-f
		RewriteCond %{REQUEST_FILENAME} !-d
		RewriteRule ^(.*)$ index.php?$1 [L,QSA]
		#RewriteRule ^(.*)$ init.ph7?$1 [L,QSA]
		
	</IfModule>

	<IfModule mod_php5.c>
		php_flag allow_url_include Off
		php_flag expose_php Off
	</IfModule>

	### Security and Spam ###

	ErrorDocument 400 /error/http/index?code=400
	ErrorDocument 401 /error/http/index?code=401
	ErrorDocument 402 /error/http/index?code=402
	ErrorDocument 403 /error/http/index?code=403
	ErrorDocument 404 /error
	ErrorDocument 405 /error/http/index?code=405
	ErrorDocument 500 /error/http/index?code=500
	ErrorDocument 501 /error/http/index?code=501
	ErrorDocument 502 /error/http/index?code=502
	ErrorDocument 504 /error/http/index?code=504
	ErrorDocument 505 /error/http/index?code=505

	# For the videos extensions
	#AddType video/ogg .ogg
	AddType video/webm .webm
	AddType video/mp4 .mp4
	AddType application/rss+xml .xml

	# Personal pH7CMS extension file
	#AddType application/x-httpd-php .ph7
	#DirectoryIndex init.ph7

**Comment by polynamaude** [comment]
Hi !

First of all, have you looked through the other already written and / or closed issue that may relate to a problem similar as yours ?

Secundo, maybe you find that the distributed .htaccess file is pretty long.
As it does lot of filtering for User-Agents that are only using bandwith and grabbing content, there's a use to such length. Most provider have much more complex htaccess / apache configuration for there server but it does reside in many files.

The exact part you removed from your .htaccess that send a 403 forbidden to many user-agents who will simply leech from your site is the one that I personnaly use in other CMS as it does a great job to save bandwidth and ressources from being overused.

May I suggest that you keep your thread as much related to what your problem may be if you want to get some help back ? There's many person who use the .htaccess "out of the box" and experience no problem at all.

Also, the least you should do is give more information about your hosting configuration such as :
Apache server version,
Log extract,
Modules enabled,
etc.

With as much "no info" as you give us, it's pretty hard to do better than speculate !

Look at the thread "mimify and static caching" as it relate somewhat to your problem (and the doubled issue you opened).

As many community based project and even more online site, there will always be some work to be done as customisation that will require in-depth technical knowledge if you want to do more than "simply point-click".

Your problem is not related to theming but to caching and the used of "curl" in the background to compress and store pages of your site ahead of serving them.

Look in your logs and you'll have many (and many more) 401 error (in your access logs).

Yes the "admin cp" is using this feature as the rest of the site as it will be much more used than on a normal CMS.

The software require a great deal of "good configuration" or "unobstructed use" of Apache as it does use many feature at they're best. Same as many years ago when Linux used to make system crash on computer with poor memory, this software pushes Apache to the limits so it can be as resource efficient as possible.

Get back with more info and we'll try to help you sort out those problem.

Also as changing your htaccess to a different version didn't solve your troubles, I'll suggest you to get back to the "standard one" as it will be easier for us to help you.

Disable "mod_security" (mod-security-crs) for the time being as it will be interfering with the troubleshooting. Also it's quite useful to know what type of hosting are you using (shared, VPS, localhost, etc).

Sincerely,
PM-Summerside

**Comment by polynamaude** [comment]
Added !
After so check-up there's no change in the latest apache 2 server (2.4.33) that is incompatible with the actual .htaccess. Could you please exlain why you make such affirmation ?

What have you tried out to make such diagnostics ?
Maybe you dislike lengthy .htaccess and would prefer to split apart but that's you own personal opinion. But talking about facts (useful stuff), I tried over apache 2.4.33, 2.4.25 (Debian Stretch), even old 2.2 of Slackware and they all go good.

A 500 server error is a client side answer, if you look into your logs you'll get what's causing this type of error and then we'll be able to have some kind of constructive discussion.
You seem to go quite fast onto conclusion without much proof.

The .htaccess file was updated last time 4 month ago https://github.com/pH7Software/pH7-Social-Dating-CMS/commit/4554eac8589d8fb332e9c485c09f69da637e077c and you're the first to make such affirmation. I have profound doubt that the problem reside somewhere else as such complaints would have came in before.

The changed made in November we're pretty small and the apache httpd server steering team doesn't make major changes in the configuration files in minor updates.

The only problem I could see have been solved in august and was related to looping on some hosting provider.

Here's the link https://github.com/pH7Software/pH7-Social-Dating-CMS/commit/dc97f5584562770765d8fca3a35963a2158f8b4c#diff-8052c42ab3b8aa06a3f5f788a4ddccc2

PM

**Comment by Foxtrek64** [comment]
I am using NearlyFreeSpeech.net as my hosting provider. It was actually their support team who suggested the incompatibility:

> The .htaccess file you are using is not compatible with current versions of Apache.
>
> You may want to consult the developer for a more recent one, preferably one that is more minimalistic for the needs of that app, and less of an editorial monoblock.
>
> Edit: The shorter version still has some extraneous (but harmless) stuff in there but does eliminate the sections with compatibility problems.

Your "bad bot" filtering actually caused NFSN's crypto bot (based off of the EFF's CertBot for Let's Encrypt) to fail (access denied, as reported by the rewrite log).

In order to get this function to work and be issued my certificates, I ended up disabling the .htaccess file entirely (renaming it to `old.htaccess`). The bot was able to function as intended at that point.

I can show my access logs and rewrite logs, but with the .htaccess changed, it's working perfectly. I will reply to the other thread and show my rewrite log for the theming issue.

I consider this issue and the theming error to be unrelated as this has to do with the .htaccess file and the other thread has to do with the theming system. They work independently of each other, and changing one has no effect on the other. Claiming they're the same issue is like me creating a ticket on my email system and my minecraft server and you telling me they're the same issue.

That said, here's the information you have requested:

Apache server version: 2.4
Log extract: 
When attempting to access the site under the default .htaccess file, the following is returned:

    [Mon Mar 26 21:42:05.007839 2018] [core:alert] [pid 86314:tid 34479619328] [client 136.63.42.213:20035] /fs7d/lfdating/public/.htaccess: Invalid command 'Deny', perhaps misspelled or defined by a module not included in the server configuration

You will see in my access file (the top section is included below), that it begins working as soon as I clean up the .htaccess file. Between the timestamps of 21:42 and 22:33, I had run the certificate issuance process with the .htaccess file disabled entirely.

Modules enabled: https://www.freebsd.org/ports/

Here's the top of my rewrite log, comments added: https://gitlab.com/snippets/1706558

**Comment by polynamaude** [comment]
At first, your "theming" issue is exactly related to the actual one as the pages don't display properly because they aren't served correctly so that they can be fetched by "curl" and cached. That's exactly why you get a good display when you disable the caching (which put a higher load on your server but doesn't change anything in the functions of the CMS).

There's a difference (really huge one) between what your "technical support" says and the actual reality. The log you just copied a line says that the "Deny" command is not supported or included in the modules your server use. This has nothing to do with Apache not supporting the actual "htaccess" that we provide but more with the fact that your provider restrict what you can include in the "htaccess" user based configuration.

It's too easy to call-on incompatibility but not to state where they're found.

In fact the "Deny" keyword is supported by Apache since the very first version and there's no backward incompatible changes that are introduced in minor version upgrade and much rarely in the major version upgrade.

I'm not sure why you sent me the link for the FreeBSD port distribution as it doesn't give me a hint about what exact modules your provider do enable.

It's quite common that you may need to disable a access control for Apache when you run a certificate creation process / bot. I need to do the same every time I run the domain validation procedure for my SSL certificates from Comodo or Verisign. You could have simply added a "Referer" validation bypass for the rules and it would have done the same. Depending on the uses of my domain, when it's as a Web server then I do exactly that as it saves me problem if I ever need a domain validation done again.

I exactly know what does EFF CertBot does and can understand that you want to have the process automated so simply add either a "Referer" or "User-Agent" rules bypass for the process and you won't have to disable the access control configuration next time.

Here's the info :

https://httpd.apache.org/docs/current/fr/mod/mod_rewrite.html

You make a rewrite condition valid for the HTTP_USER_AGENT and HTTP_REFERER used by EFF Crypto Bot (you can find those information in your access log file).

It's quite possible that your provider has security modules enabled that may interfere with the CMS. As I told you in my last message, I would recommend that you check out with them if they use mod_security and in such case to have it disabled. One of the reason we have our own security configuration is that the CMS does a great job on it's own.

I'll end by suggesting that you first install a version on your own PC at home and do a installation from a "local". It will allow you to get used with the software and do easier troubleshooting.

I'll also suggest to make it easier troubleshooting to try the following as a .htaccess

--- start --

RewriteEngine On
RewriteBase /
SetEnv HTTP_MOD_REWRITE On
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^(.*)$ index.php?$1 [L,QSA]
php_flag allow_url_include Off
php_flag expose_php Off

-- end --

You don't have to use the <If> </> division when you're sure that the modules are enabled and by assuming that they are (like I'm doing in this file). You make sure that the required modules are effectively loader.

I looked at the answer you gave me in the other thread and it does exactly confirm what my belief was and that I stated at first. It's no something as a "theming" issue but has to do with the files not getting delivered. As they can't be cached, they can't be delivered so you get a messed up display.

Good luck !

**Comment by polynamaude** [comment]
I'm not sure if the latest version are on SourceForge. Anyway, the standard update would suggest that you use a git so maybe to start at first make a git clone, branch the latest version and use that one to get the site running.

I noticed by looking at you logs that you seem to be trying out the software on a live public server ? I'd suggest you get used to the product before going into a "production".

Don't want to disappoint you, but this software is pretty easy to get using but you'll need to do some real work of customising if you want to get something useful out of it and it's not by stopping at a "doesn't work" that you'll go really far.

Take a look inside at the source code to get some in-deep understanding of they way stuff is done. No real work is easy done and I can guarantee you that it won't go far if you have in mind that everything will only be a point-and-click operation.

**Comment by pH-7** [comment]
Hi all!

1. The version on Sourceforge (v12.1.2) is the latest one. Only really minor changes have been pushed on Github since the last release.

2- According to your log, it looks like your are running it with Apache 2.4+. Apparently, in your case, "mod_access_compat" isn't enabled, isn't it? 
I'm working on a fix to make the apache files compatible with either 2.2 or 2.4 versions (even if mod_access_compat hasn't been enabled). It should be published very soon.

Thanks!

**Comment by polynamaude** [comment]
Hi all !
Thanks @pH-7 for the confirmation that Sourceforge is up to date with the Git. I wasn't sure so I suggested to use a repo as it's also more convenient when you want to try some code modification. You make a branch and you trash it later-on when you're finished with trying your coding.

I've made some test and contacted different hosting manager (Namecheap, Godaddy and iWeb plus some local ones). They all confirmed my first thought as any htaccess rule that do a minimum of restriction against anything except "browsing user" won't allow automatic certificate validation (SSL because they need mainly to have direct access to the server and aren't user-agents normally included in the standard "browsing user".

So SSL certificate will always be possible on a server running pH7CMS, but some little tricks will be required if you want to use something like the cPanel or Plex modules that do automatic validation and generation once expiry. For example the Namecheap SSL plugin on cPanel for user using Namecheap with Positive SSL certificate. You have either to :
1 - send a tickets to ask the agents to validate for you 
2 - to manually modify your script to accept the SSL bot that will be calling upon your server and requesting a specific file (normally a long filename ending by .txt).
3 - use another validation method (for example a modified DNS record).
4 - or be creative as it's not normally something that is done really often to validate your domain for SSL certificate.

P.S. I love complex Apache file as they're much faster than any other way of managing user access in the way of either browser, time of day, IP, etc... check https://github.com/polynamaude/A2-BadIPs
Maybe we could integrate some kind of IP validation later-on as it can be used easily with the actual DDoS. I use the dbm file format as they other type of file produced are quite large and only for demonstration. The documentation is lacking but I promise a README in the next two days. It was a "emergency" coding so it's far from neat and clean.

Thanks
